### ISSUE TITLE
client hangs on raspberry pi with high load

### ISSUE THREAD

**Issue by quantenschaum** [badaix/snapcast#120]
I just installed this great software (client) on a raspberry pi B running minibian jessie. After some time the playback stops, but the client is still shown in the android app. The client produces 100% CPU load. I have to `systemctl restart snapclient`, which takes some time. 

`systemctl status snapclient` tells me

```
● snapclient.service - Snapcast client
   Loaded: loaded (/lib/systemd/system/snapclient.service; enabled)
   Active: active (running) since Thu 2016-10-27 17:37:37 BST; 14min ago
  Process: 868 ExecStart=/usr/sbin/snapclient $SNAPCLIENT_OPTS (code=exited, status=0/SUCCESS)
 Main PID: 869 (snapclient)
   CGroup: /system.slice/snapclient.service
           └─869 /usr/sbin/snapclient -d -h nas -s dmix:CARD=Device,DEV=0

Oct 27 17:37:37 kitchen systemd[1]: PID file /var/run/snapclient.pid not readable (yet?) after start.
Oct 27 17:37:37 kitchen snapclient[869]: daemon started
Oct 27 17:37:37 kitchen systemd[1]: Started Snapcast client.
Oct 27 17:37:38 kitchen snapclient[869]: Connected to 192.168.222.10
Oct 27 17:47:13 kitchen snapclient[869]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_UNPARSEABLE_STREAM
Oct 27 17:47:13 kitchen snapclient[869]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
```

**Comment by badaix** [comment]
is this reproducible?

**Comment by 1337sup3rh4x0r** [comment]
How die you create the pipe sink? This might be related to another closed issue around here where snapcast was writing to a file instead of a pipe and at some point this file gets too big. Just an idea worth checking

**Comment by badaix** [comment]
Yikes, I have the same error on OpenWrt just now:

this is what `logread` shows:
````
Thu Nov 10 07:26:44 2016 daemon.warn dnsmasq-dhcp[1194]: no address range available for DHCP request via wlan0
Thu Nov 10 07:54:39 2016 daemon.err snapclient[1158]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Thu Nov 10 07:54:39 2016 daemon.err snapclient[1158]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
````

along with high CPU usage:
````
Mem: 25932K used, 2648K free, 88K shrd, 1816K buff, 5356K cached
CPU:  99% usr   0% sys   0% nic   0% idle   0% io   0% irq   0% sirq
Load average: 1.00 1.01 0.86 3/42 3410
  PID  PPID USER     STAT   VSZ %VSZ %CPU COMMAND
 1158     1 root     S <  14724  51%  99% /usr/bin/snapclient -d
````

Seems that in case of a FLAC LOST SYNC (for whatever reason), the Snapclient should deal with it in some smarter way.

**Comment by quantenschaum** [comment]
Yes, it is reproducible. My workaround currently is to use the ogg codec.

**Comment by badaix** [comment]
And again... this time on RPI:
`````
Nov 10 12:42:30 schlafzimmer snapclient[645]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_UNPARSEABLE_STREAM
Nov 10 12:42:30 schlafzimmer snapclient[645]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Nov 10 13:17:27 schlafzimmer snapclient[657]: daemon started
`````
@quantenschaum what version do you use? I see this now for the first time, right after switching my "production" server to the spotify branch.

**Comment by quantenschaum** [comment]
I am using the current release `snapclient v0.9.0`

**Comment by badaix** [comment]
prebuild `dpk`? Same for server?

**Comment by quantenschaum** [comment]
Yes, I download both, server and client, from the releases page and installed the `deb`. The client on the Pi, the server on an Ubuntu machine, both v0.9.0.

**Comment by badaix** [comment]
@quantenschaum can you please test if the attached 0.9.90 solves the problem for you?
It smells like a server issue. I could reproduce it with a development version, but not with the released 0.9.0 (different threads were sending simultaneously over a non-thread safe socket).

[snapclient_0.9.90_amd64.deb.gz](https://github.com/badaix/snapcast/files/589900/snapclient_0.9.90_amd64.deb.gz)
[snapclient_0.9.90_armhf.deb.gz](https://github.com/badaix/snapcast/files/589901/snapclient_0.9.90_armhf.deb.gz)
[snapserver_0.9.90_amd64.deb.gz](https://github.com/badaix/snapcast/files/589902/snapserver_0.9.90_amd64.deb.gz)
[snapserver_0.9.90_armhf.deb.gz](https://github.com/badaix/snapcast/files/589903/snapserver_0.9.90_armhf.deb.gz)

**Comment by quantenschaum** [comment]
Thanks! I will test it.

**Comment by quantenschaum** [comment]
The issue persists. After some time the playback stops, snapclient hangs with high cpu usage and the log says
```
Nov 15 18:35:33 kitchen snapclient[18013]: 2016-11-15 18-35-33 [out] Chunk: 8#0119#0118#0117#011255#011126
Nov 15 18:35:34 kitchen snapclient[18013]: 2016-11-15 18-35-34 [out] Chunk: 9#0119#0118#0117#011289#011126
Nov 15 18:35:35 kitchen snapclient[18013]: 2016-11-15 18-35-35 [out] Chunk: 10#01110#0119#0118#011322#011126
Nov 15 18:35:35 kitchen snapclient[18013]: 2016-11-15 18-35-35 [Err] Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Nov 15 18:35:35 kitchen snapclient[18013]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Nov 15 18:35:35 kitchen snapclient[18013]: Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Nov 15 18:35:35 kitchen snapclient[18013]: 2016-11-15 18-35-35 [Err] Got error callback: FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC
Nov 15 18:35:36 kitchen snapclient[18013]: 2016-11-15 18-35-36 [out] Chunk: 10#01110#0119#0118#011355#011126
Nov 15 18:35:36 kitchen snapclient[18013]: 2016-11-15 18-35-36 [out] Failed to get chunk
Nov 15 18:35:41 kitchen snapclient[18013]: 2016-11-15 18-35-41 [out] time sync main loop
Nov 15 18:35:41 kitchen snapclient[18013]: 2016-11-15 18-35-41 [out] No chunk received for 5000ms. Closing ALSA.
Nov 15 18:35:46 kitchen snapclient[18013]: 2016-11-15 18-35-46 [out] time sync main loop
Nov 15 18:35:51 kitchen snapclient[18013]: 2016-11-15 18-35-51 [out] time sync main loop
Nov 15 18:35:56 kitchen snapclient[18013]: 2016-11-15 18-35-56 [out] time sync main loop
Nov 15 18:36:01 kitchen snapclient[18013]: 2016-11-15 18-36-01 [out] time sync main loop
```

It seems to continue to run, but does not restart the playback. It seems as if it does not recover from an error in the flac stream. It stream over wifi, so some the stream might loose packets (despite using TCP). It should just restart on error, but it hangs. BTW, why do you use TCP not UPD for streaming?

**Comment by badaix** [comment]
In my case the decoding was stuck in an infinite loop, caused by:

```C++
void Controller::onMessageReceived(ClientConnection* connection, const msg::BaseMessage& baseMessage, char* buffer)
{
...
	pcmChunk->deserialize(baseMessage, buffer);
	if (decoder_->decode(pcmChunk))
...
}
```

```C++
bool FlacDecoder::decode(msg::PcmChunk* chunk)
{
	std::lock_guard<std::mutex> lock(mutex_);
...
	if (!FLAC__stream_decoder_process_single(decoder))
...
}
```

the `FLAC__stream_decoder_process_single` called under the hood `read_callback` with `flacChunk->payloadSize == 0` => `bytes == 0`:

```C++
FLAC__StreamDecoderReadStatus read_callback(const FLAC__StreamDecoder *decoder, FLAC__byte buffer[], size_t *bytes, void *client_data)
{
	if (flacChunk != NULL)
	{
		static_cast<FlacDecoder*>(client_data)->cacheInfo_.isCachedChunk_ = false;
		if (*bytes > flacChunk->payloadSize)
			*bytes = flacChunk->payloadSize;

		memcpy(buffer, flacChunk->payload, *bytes);
		memmove(flacChunk->payload, flacChunk->payload + *bytes, flacChunk->payloadSize - *bytes);
		flacChunk->payloadSize = flacChunk->payloadSize - *bytes;
		flacChunk->payload = (char*)realloc(flacChunk->payload, flacChunk->payloadSize);
	}
	return FLAC__STREAM_DECODER_READ_STATUS_CONTINUE;
}
```

by "construction" the server sends only fully decodable FLAC chunks in each `PcmChunk` message.
I decided to use TCP to avoid situations like this. On the wire (or in the air) such chunks might be split into several IP packets. The decoder needs the chunks to be complete (all packets, in correct order) so that they can be properly decoded. TCP takes care that no packet is lost and all are ordered.

This also means: if the chunks are not properly encoded, all connected clients should receive the same (on bit level) PcmChunks and run at the same time into the infinite loop. In an interims version (unreleased dev version) the sending of the PcmChunks was not protected by a mutex (in `bool StreamSession::send(const msg::BaseMessage* message) const`). Without the mutex I was able to reproduce this decoder loop, because timesync messages were sent simultaneously from a different thread.

Of course should the FLAC decoder not run into a loop in case of an undecodable chunk. But more important: there shouldn't be any non-decodable chunks.

I'm also wondering why this bug has not been reported earlier, i.e. what are you doing differently ;).
My test-setup consists of several Raspberry clients (most of them Raspbian Jessie, one Arch), Ubuntu x64 client, Mac client, OpenWRT MIPS client and an Ubuntu x64 server. I have an mpd stream, a Spotify stream and an AirPlay stream running in parallel. 
Without mutex => after 5min at least one client will loop. With mutex: no problem at all.

Might it be related to minibian?

**Comment by quantenschaum** [comment]
Ok, how does TCP loose data? It is designed to not to loose data, but if the connection is interrupted, then it will loose data anyway. So the loop must be avoided and it should recover from it. But what is going on here? I don't know. Might it be related to streaming over wifi? I can try again using a wired connection. Minibian is basically Jessie, just stripped down, but can apt-get anything you need.

**Comment by badaix** [comment]
if a single packet would be dropped then also ogg should stop working. 
The messages in the "Snapcast protocol" are separated by their length. The client reads first the message header (a fixed amount of bytes), followed by reading "message size" bytes (the size is part of the header). As soon as a single byte would be dropped, the stream would be out of sync and the client would not be able to continue deserializing the stream into single messages.
In other words: I have no idea what's going on here :confused:

**Comment by badaix** [comment]
I now have a Minibian running for two hours (over WiFi). No problems so far. 
I think I will for debugging add the possibility to write the received data stream to a file and to "replay" this file later on.

**Comment by badaix** [comment]
After 14h Minibian lost WiFi connection. Till then there were no problems.

**Comment by quantenschaum** [comment]
Hi, I replaced the USB WiFi adapter on the Pi and now it does not hang when streaming with flac. This was the problem I guess. But the infinite loop is not that good anyway. Great work, thank you for this software. I was looking for Linux based in-sync multiroom audio for quite a while. I'd like to buy you a beer or something like that...
